src/librustc_trans/cabi_mips64.rs

```diff
@@ -8,50 +8,153 @@
+        if let layout::Int(i, signed) = scalar.value {
+            if i.size().bits() < bits {
+                if let PassMode::Direct(ref mut attrs) = arg.mode {
+                    attrs.set(if signed || i.size().bits() >= 32 {
+                        ArgAttribute::SExt
+                    } else {
+                        ArgAttribute::ZExt
+                    });
+                }
+            }
+        }
+    }
+}
+
+fn bits_to_int_reg(bits: u64) -> Reg {
+    if bits <= 8 {
+        Reg::i8()
+    } else if bits <= 16 {
+        Reg::i16()
+    } else if bits <= 32 {
+        Reg::i32()
+    } else {
+        Reg::i64()
+    }
+}
 
-fn classify_ret_ty<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>,
-                             ret: &mut ArgType<'tcx>,
-                             offset: &mut Size) {
+fn classify_ret_ty<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>, ret: &mut ArgType<'tcx>) {
     if !ret.layout.is_aggregate() {
-        ret.extend_integer_width_to(64);
+        extend_integer_width_mips(ret, 64);
+        return;
+    }
+
+    let size = ret.layout.size;
+    let bits = size.bits();
+    if bits <= 128 {
+        // Unlike other architectures which return aggregates in registers, MIPS n64 limits the
+        // use of float registers to structures (not unions) containing exactly one or two
+        // float fields.
+
+        if let layout::FieldPlacement::Arbitrary { .. } = ret.layout.fields {
+            let float_regs: Vec<Reg> =
+                (0..ret.layout.fields.count()).flat_map(|i| {
+                    match ret.layout.field(cx, i).abi {
+                        layout::Abi::Scalar(ref scalar) => match scalar.value {
+                            layout::F32 => Some(Reg::f32()),
+                            layout::F64 => Some(Reg::f64()),
+                            _ => None
+                        },
+                        _ => None
+                    }
+                }).collect();
+
+            if ret.layout.fields.count() == float_regs.len() {
+                if float_regs.len() == 1 {
+                    ret.cast_to(Uniform { unit: float_regs[0], total: float_regs[0].size });
+                    return;
+                } else if float_regs.len() == 2 {
+                    ret.cast_to(CastTarget::Pair(float_regs[0], float_regs[1]));
+                    return;
+                }
+            }
+        }
+
+        // Cast to a uniform int structure
+        ret.cast_to(Uniform {
+            unit: bits_to_int_reg(bits),
+            total: size
+        });
     } else {
         ret.make_indirect();
-        *offset += cx.tcx.data_layout.pointer_size;
     }
 }
 
-fn classify_arg_ty(cx: &CodegenCx, arg: &mut ArgType, offset: &mut Size) {
+fn classify_arg_ty<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>, arg: &mut ArgType<'tcx>) {
     let dl = &cx.tcx.data_layout;
     let size = arg.layout.size;
-    let align = arg.layout.align.max(dl.i32_align).min(dl.i64_align);
 
     if arg.layout.is_aggregate() {
```

Comment:
Should handle the simple (non-aggregate) case first with an early return, to avoid the indentation in the aggregate case.

Reply:
Fixed